### PR TITLE
Patterns: Fix patterns that could be copied when searching

### PR DESCRIPTION
**Pull request by stephanethomas**
This pull request seeks to address https://github.com/Automattic/dotcom-forge/issues/6323 by enforcing a limit of three patterns per category that can be copied while being logged out. This fixes the case where it would be possible to copy a locked pattern by looking it up:

image

#### Testing instructions

1. Run `git checkout lock/patterns` and start your server, or access a [live branch](https://github.com/Automattic/wp-calypso/pull/89023#issuecomment-2025796292)
2. Make sure you are logged out from WordPress.com
3. Open the [`Patterns` page](http://calypso.localhost:3000/patterns?grid=1)
4. Search for `contact`
5. Assert that the pattern that can be copied are shown at the top
6. Open the [`Footers Patterns` page](http://calypso.localhost:3000/patterns/footer?grid=1)
7. Assert that only the first three patterns can be copied
8. Search for `Footer: Simple, contact info left` now
9. Assert that the pattern displayed can be copied
10. Search for `Four columns of menus with social`
11. Assert that the pattern displayed cannot be copied
12. Navigate to the [`About Page Patterns` page](http://calypso.localhost:3000/patterns/layouts/about?grid=1)
13. Assert that only the first three patterns can be copied

### CODE REVIEW

**Review by fredrikekelund — COMMENTED** [review]
As I mentioned in D143680-code, the other option I would consider is to update `client/my-sites/patterns/hooks/use-patterns.ts` like so:

```ts
export function getPatternsQueryOptions(
	locale: string,
	category: string,
	queryOptions: Omit< UseQueryOptions< Pattern[] >, 'queryKey' > = {}
): UseQueryOptions< Pattern[] > {
	return {
		queryKey: [ 'pattern-library', 'patterns', locale, category ],
		queryFn() {
			return wpcom.req.get( `/ptk/patterns/${ locale }`, {
				categories: category,
				post_type: 'wp_block',
			} );
		},
		staleTime: 5 * 60 * 1000,
		...queryOptions,
		select( patterns ) {
			const categories: Record< string, number > = {};

			for ( const pattern of patterns ) {
				for ( const categoryName of Object.keys( pattern.categories ) ) {
					categories[ categoryName ] = ( categories[ categoryName ] ?? 0 ) + 1;
					pattern.can_be_copied_without_account = categories[ categoryName ] <= 3;
				}
			}

			if ( queryOptions.select ) {
				return queryOptions.select( patterns );
			}

			return patterns;
		},
	};
}
```

And `client/my-sites/patterns/types.ts` like so:

```ts
import type { Pattern as AssemblerPattern } from 'calypso/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/types';

export type Pattern = AssemblerPattern & {
	can_be_copied_without_account?: boolean;
};
```

The upside to this approach, as I see it, is that `can_be_copied_without_account` is only relevant to the Pattern Library (not to other consumers of `/ptk/patterns`) and, overall, that property seems like a front-endy concern.

Do you have a preference between these two approaches, @stephanethomas?

**Comment by stephanethomas** [comment]
For the records we discussed that over Slack, and decided to follow what Fredrik suggested.

File: client/my-sites/patterns/hooks/use-patterns.ts [hunk 1]
```diff
@@ -17,6 +27,37 @@ export function getPatternsQueryOptions(
 		},
 		staleTime: 5 * 60 * 1000,
 		...queryOptions,
+		select( patterns: Pattern[] ) {
+			const patternCounts: PatternCounts = {
+				page: {},
+				regular: {},
```
Comment (fredrikekelund) [inline client/my-sites/patterns/hooks/use-patterns.ts]:
```suggestion
				[ CATEGORY_PAGE ]: {},
				[ CATEGORY_REGULAR ]: {},
```
Should we be using those constants there as well?

File: client/my-sites/patterns/hooks/use-patterns.ts [hunk 2]
```diff
@@ -17,6 +27,37 @@ export function getPatternsQueryOptions(
...
+			};
+
+			// Limits the number of patterns that can be copied to the first three patterns from each category
+			for ( const pattern of patterns ) {
+				const categoryNames = Object.keys( pattern.categories );
+
+				const patternType = categoryNames.includes( CATEGORY_PAGE ) ? 'page' : 'regular';
```
Comment (fredrikekelund) [inline client/my-sites/patterns/hooks/use-patterns.ts]:
```suggestion
				const patternType = categoryNames.includes( CATEGORY_PAGE ) ? CATEGORY_PAGE : CATEGORY_REGULAR;
```

**Review by fredrikekelund — COMMENTED** [review]
Overall this looks good and seems to work as expected 👍 

While this is a more technically correct experience compared to what we currently have for search, it might still seem a bit weird to users that, for example, the first pattern in the search results can't be copied, the next two can, then not, then two more that can be copied again.

It makes me think that maybe we should sort search results by the `can_be_copied_without_account`. WDYT, @stephanethomas?

File: client/my-sites/patterns/hooks/use-patterns.ts (hunk 1 above)
Comment (stephanethomas) [reply client/my-sites/patterns/hooks/use-patterns.ts]: I thought about that as well but was hesitant to follow that route: these are not categories per se but more pattern types. However we could use the `PAGES` and `REGULAR` [types](https://github.com/Automattic/wp-calypso/blob/2cc0abfd45d41b68a63afff7c63949e1e31be8e0/client/my-sites/patterns/types.ts#L18-L21) for that, I'll update that code.

**Comment by stephanethomas** [comment]
> It makes me think that maybe we should sort search results by the `can_be_copied_without_account`.

That is a good idea, I've updated this pull request accordingly.

**Review by fredrikekelund — APPROVED** [review]
Looks good and tests well 👍

**Pull request merged**